crates/rslint_parser/src/syntax/stmt.rs

```diff
@@ -94,68 +94,77 @@ pub(super) fn is_semi(p: &Parser, offset: usize) -> bool {
-		T![for] => parse_for_statement(p).ok().unwrap(),
-		T![do] => parse_do_statement(p).ok().unwrap(),
-		T![switch] => parse_switch_statement(p).ok().unwrap(),
-		T![try] => parse_try_statement(p).ok().unwrap(), // it is only ever Err if there's no try
-		T![return] => parse_return_statement(p).ok().unwrap(),
-		T![break] => parse_break_statement(p).ok().unwrap(),
-		T![continue] => parse_continue_statement(p).ok().unwrap(), // It is only ever Err if there's no continue keyword
-		T![throw] => parse_throw_statement(p).ok().unwrap(),
-		T![debugger] => parse_debugger_statement(p).ok().unwrap(),
-		T![function] => function_declaration(p),
-		T![class] => class_declaration(p),
-		T![ident]
-			if p.cur_src() == "async"
-				&& p.nth_at(1, T![function])
-				&& !p.has_linebreak_before_n(1) =>
-		{
-			function_declaration(p)
-		}
+		T![for] => parse_for_statement(p),
+		T![do] => parse_do_statement(p),
+		T![switch] => parse_switch_statement(p),
+		T![try] => parse_try_statement(p), // it is only ever Err if there's no try
+		T![return] => parse_return_statement(p),
+		T![break] => parse_break_statement(p),
+		T![continue] => parse_continue_statement(p), // It is only ever Err if there's no continue keyword
+		T![throw] => parse_throw_statement(p),
+		T![debugger] => parse_debugger_statement(p),
+		T![function] => parse_function_declaration(p),
+		T![class] => parse_class_declaration(p),
+		T![ident] if is_at_async_function(p, true) => parse_function_declaration(p),
 
 		T![ident] if p.cur_src() == "let" && FOLLOWS_LET.contains(p.nth(1)) => {
 			variable_declaration_statement(p)
 		}
 		// TODO: handle `<T>() => {};` with less of a hack
-		_ if p.at_ts(STARTS_EXPR) || p.at(T![<]) => return expr_statement(p),
-		_ => {
-			let err = p
-				.err_builder("Expected a statement or declaration, but found none")
-				.primary(
-					p.cur_tok().range,
-					"Expected a statement or declaration here",
-				);
+		_ if p.at_ts(STARTS_EXPR) || p.at(T![<]) => parse_expression_statement(p),
+		_ => Absent,
+	};
 
+	match res {
+		Absent => {
 			// We must explicitly handle this case or else infinite recursion can happen
 			if p.at_ts(token_set![T!['}'], T![import], T![export]]) {
+				let err = p
+					.err_builder("Expected a statement or declaration, but found none")
+					.primary(
+						p.cur_tok().range,
+						"Expected a statement or declaration here",
+					);
 				p.err_and_bump(err, JS_UNKNOWN_STATEMENT);
-				return None;
 			}
-			SingleTokenParseRecovery::with_error(
-				recovery_set.into().unwrap_or(STMT_RECOVERY_SET),
-				JS_UNKNOWN_STATEMENT,
-				err,
-			)
-			.recover(p);
-			return None;
-		}
-	};
+			let recovered = res.or_recover(
+				p,
+				ParseRecovery::new(
+					JS_UNKNOWN_STATEMENT,
+					recovery_set.into().unwrap_or(STMT_RECOVERY_SET),
+				),
+				js_parse_error::expected_statement,
+			);
 
-	Some(res)
+			if recovered.is_ok() {
+				recovered.ok()
+			} else {
+				None
+			}
+		}
```

Comment:
You can simply use `recovered.ok()`

```suggestion
			recovered.ok(
		}
```